File: LeetCode/[M] - 735. Asteroid Collision/solution.go
package main

func asteroidCollision(asteroids []int) []int {
	stack := make([]int, 0)
	for _, a := range asteroids {
		for len(stack) != 0 && (stack[len(stack)-1] > 0 && a < 0) && (stack[len(stack)-1] < -a) {
			stack = stack[:len(stack)-1]
		}
		if len(stack) == 0 || !(stack[len(stack)-1] > 0 && a < 0) {
			stack = append(stack, a)
		} else if stack[len(stack)-1] == -a {
			stack = stack[:len(stack)-1]
		}
	}
	return stack
}

File: LeetCode/[M] - 735. Asteroid Collision/solution.py
class Solution:
    def asteroidCollision(self, asteroids: list[int]) -> list[int]:
        stack = []
        for ast in asteroids:
            while stack and (stack[-1] > 0 and ast < 0) and stack[-1] < -ast:
                stack.pop()
            if not stack or not (stack[-1] > 0 and ast < 0):
                stack.append(ast)
            elif stack[-1] == -ast:
                stack.pop()
        return stack
